### ISSUE TITLE
Cheaters (Macro Players)

### ISSUE THREAD

**Issue by qwe812** [haxball/haxball-issues#142]
Hi Basro, can you do something about this? Like "kicks per second limit" or etc.

**Comment by loveuhaxball** [comment]
yea this big problem. i hate macro players

**Comment by Kozlov98** [comment]
Yes, i hate macro players too :(

**Comment by basro** [comment]
An optional rate limit for kicks is a possibility.

**Comment by Leo-MapCreator** [comment]
I dont have the numbers right now, but i can make a macro and test it for you guys, if im not mistaken, they kick thrice in around 100 ms, for every time they press the key once

**Comment by hypehaxx** [comment]
Hey @basro! Limit for kicks is really bad idea, because one of the most important things in "professional" haxball is "battle spam" skill. If you can't spam good in the 1x1 episodes in 4x4 game - you are useless player for all team. You can just look at latest matches in Nations Cup just to see how it works. For example:  https://youtu.be/CfJzM9l9QEI?t=526

**Comment by basro** [comment]
@hypehaxx, I said optional, meaning the room host would decide if he wants to apply a kick rate limit.
If I implemented such a thing, the default setting would be same as haxball is today.

**Comment by Niko0304** [comment]
@hypehaxx 
Yep, it's truth. We had kicks (clicks) that were naturally limited to 16,(6) ms in the 60 fps game (as 1 frame [16,(6) ms] per click held, 1 frame per click distance) and it works.
I don't assume how it works now (in the html5 edition), when we have 144 fps with 144hz monitors, 240 with 240 ones, etc. But I clearly believe, that @basro should not put a more serious restriction than 16,(6) or 33,(3) ms, because another ones can totally kill the game physics.

**Comment by basro** [comment]
@Niko0304 the game physics still run at 60hz so the 16ms hard limit is still there. HTML5 HaxBall implements interpolation to achieve the smoother animation in 144hz monitors.

**Comment by haxballer** [comment]
Removing macro players is a great idea, but you have to think not to limit the speed of pressing x, if it is possible to intelligently block auto-clicks, would be great!

**Comment by helppe** [comment]
@Leo-MapCreator 3 kicks in 1 second is easily done during a match without macro, maybe 8 in 1 second is more reasonable because harder to do and not too high to detect slower macros. 
if you look the macro users you see that there are really a lot of kicks in 1 second.

**Comment by Leo-MapCreator** [comment]
@basro @helppe I got the numbers, its 4 kicks in 35ms, and a wait time of 5ms to run again the key, and 2 kicks in 45ms for slower macros, with a wait time of 300ms to use the key again

For slower macros, its 4.44 kicks in 100ms, and for the faster its 11.43 kicks in 100ms

A normal person can kick around 0.8 kicks in 100ms, but you can get from 1 kick in 100ms up to 1.5/100ms if giving all you can
[Source](https://scratch.mit.edu/projects/27310194/): https://gyazo.com/003d99868394586b97eeebc474a2ebdb (this was mine and I kick really fast compared to other haxballers)
Check the comments for more kicks/limit per persons, the highest was 1.69 in 100ms


With that in mind, I think that the kick limiter should be around 2 to 3 kicks in 100ms

**Comment by Leo-MapCreator** [comment]
@basro Having in mind that the game run at 60Hz, so the 16ms is there, i propose this two setups for the kick limiter:
1 kick in 48ms, meaning you can kick twice in 96ms (almost 100ms) [20,83 kicks/second]  
1 kick in 32ms, meaning you can kick thrice in 96ms (almost 100ms) [31,25 kicks/second]
